src/manifolds/GeneralizedStiefel.jl

```diff
@@ -0,0 +1,242 @@
+        )
+    end
+    if any(size(X) != representation_size(M))
+        return DomainError(
+            size(X),
+            "The matrix $(X) does not lie in the tangent space of $(p) on the Generalized Stiefel manifold of dimension ($(n),$(k)), since its dimensions are wrong.",
+        )
+    end
+    if !isapprox(p' * M.B * X + X' * M.B * p, zeros(k, k); kwargs...)
+        return DomainError(
+            norm(p' * M.B * X + X' * M.B * p),
+            "The matrix $(X) does not lie in the tangent space of $(p) on the Generalized Stiefel manifold of dimension ($(n),$(k)), since x'Bv + v'Bx is not the zero matrix.",
+        )
+    end
+end
+
+embed!(::GeneralizedStiefel, q, p) = (q .= p)
+
+embed!(::GeneralizedStiefel, Y, p, X) = (Y .= X)
+
+@doc doc"""
+    inner(M::GeneralizedStiefel, p, X, Y)
+
+Compute the inner product for two tangent vectors `X`, `Y` from the tangent space of `p`
+on the [`GeneralizedStiefel`](@ref) manifold `M`. The formula reads
+
+````math
+(X, Y)_p = \operatorname{trace}(v^{\mathrm{H}}Bw),
+````
+i.e. the metric induced by the scalar product `B` from the embedding, restricted to the
+tangent space.
+"""
+inner(M::GeneralizedStiefel, p, X, Y) = dot(X, M.B * Y)
+
+
+@doc doc"""
+    manifold_dimension(M::GeneralizedStiefel)
+
+Return the dimension of the [`GeneralizedStiefel`](@ref) manifold `M`=$\operatorname{St}(n,k,B,𝔽)$.
+The dimension is given by
+
+````math
+\begin{aligned}
+\dim \mathrm{St}(n, k, B, ℝ) &= nk - \frac{1}{2}k(k+1) \\
+\dim \mathrm{St}(n, k, B, ℂ) &= 2nk - k^2\\
+\dim \mathrm{St}(n, k, B, ℍ) &= 4nk - k(2k-1)
+\end{aligned}
+````
+"""
+manifold_dimension(::GeneralizedStiefel{n,k,ℝ}) where {n,k} = n * k - div(k * (k + 1), 2)
+manifold_dimension(::GeneralizedStiefel{n,k,ℂ}) where {n,k} = 2 * n * k - k * k
+manifold_dimension(::GeneralizedStiefel{n,k,ℍ}) where {n,k} = 4 * n * k - k * (2k - 1)
+
+@doc doc"""
+    project_point(M::GeneralizedStiefel,p)
+
+Projects `p` from the embedding onto the [`GeneralizedStiefel`](@ref) `M`, i.e. compute `q`
+as the polar decomposition of $p$ such that $q^{\mathrm{H}}Bq$ is the identity,
+where $\cdot^{\mathrm{H}}$ denotes the hermitian, i.e. complex conjugate transposed.
+"""
+project_point(::GeneralizedStiefel, ::Any...)
+
+function project_point!(M::GeneralizedStiefel, q, p)
+    s = svd(p)
+    e=eigen(s.U' * M.B * s.U)
+    qsinv = e.vectors * Diagonal(1 ./ sqrt.(e.values))
+    q .= s.U * qsinv * e.vectors' * s.V'
+    return q
+end
+
+@doc doc"""
+    project_tangent(M, p, X)
+
+Project `X` onto the tangent space of `p` to the [`GeneralizedStiefel`](@ref) manifold `M`.
+The formula reads
+
+````math
+\operatorname{proj}_{\operatorname{St}(n,k)}(p,X) = X - p\operatorname{Sym}(p^{\mathrm{H}}BX),
+````
+
+where $\operatorname{Sym}(y)$ is the symmetrization of $y$, e.g. by
+$\operatorname{Sym}(y) = \frac{y^{\mathrm{H}}+y}{2}$.
```

Comment:
The implementation does something different though (IIRC it just takes the upper triangular part of the matrix).

Reply:
Right, I will change this line.